### PR TITLE
Changed formula for PPM SIT charge

### PR DESCRIPTION
**Pull request by reggieriser**
## Description

This PR changes the current SIT charge calculation to use a new formula:

> (185A SIT first day rate * CWT) +
> (185B SIT additional day rate * additional days * CWT) +
> 210A SIT PD 30 miles or less for SIT PD schedule of service area +
> 225A SIT PD Self/Mini Storage for services schedule of service area

This formula applies to PPMs only.  HHGs use a different formula (and a baseline minimum weight of 1000 pounds); that will be implemented in a future story.  Also, note that this total will be reduced by the appropriate discount SIT rate found in the `transportation_service_provider_performance` table.  The 185A, 185B, and 210A parts use the SIT discount rate; the 225A part uses the linehaul discount rate.

Since this calculation is part of the rate engine, quite a few tests that ultimately use the rate engine had to be adjusted to add 210A and 225A rates and change expected amounts.

## Reviewer Notes

I'm not 100% sure if the 225A rate should be based on the service area's SIT PD schedule or the services schedule (the schedule numbers do differ for some service areas).  I have posed the question to the govies here: https://defensedigitalservice.slack.com/archives/G9CQY38J2/p1550253015007100

**Update 1**:
The response back was that 225A should be based on the services schedule, but this has led to additional questions about using 225A with SIT.  Will update this PR as needed.

**Update 2:**
We have the correct formula for PPM SIT, but our initial assumption that it also applied to HHG was incorrect.  I am rolling back the HHG calculation part of this PR to what previously existed and leave the HHG SIT changes for a future story when we build out the Office SIT panel.

**Update 3:**
There is some additional clarification needed regarding the discount rates to apply for the parts of the PPM SIT charge.  I am putting this PR in WIP status until we get further info.  See https://defensedigitalservice.slack.com/archives/G8HJRJBMZ/p1550784590044900 for context.

**Update 4:**
The 185A, 185B, and 210A parts use the SIT discount rate; the 225A part uses the linehaul discount rate.

## Setup

Run `db_dev_e2e_populate` and start up the office UI.  Pick a PPM shipment and use the Storage Calculator to determine the maximum SIT obligation.  Compare this to the amount calculated by hand from the formula above (don't forget the discount rate!).

For instance, for move with locator `VGHEIS` and a weight of 5000 pounds and 5 days of SIT, the calculation should look like:

> (1717 * 50) +
> (74 * 4 * 50) +
> 244200 +
> 31100

That equals `375950` cents before any discount rates are applied.  The first three parts should use the SIT discount rate, and the last part should use the linehaul discount rate.  Given the source rate area `US6`, the destination region `8`, and the code of service `D`, the SIT discount rate is `0.280399758368731` and the linehaul discount rate is `0.661429312080145`.

So, the SIT discount part is:

> (1717 * 50) +
> (74 * 4 * 50) +
> 244200

That equals `344850` cents which we then apply the SIT discount to result in `248154` cents.

The linehaul discount part is `31100` which we then apply the linehaul discount to result in `10530` cents.

Add those together and you get `258684` cents, or `$2586.84`.

There are also server tests that you can run for validating the calculations.

## Code Review Verification Steps

* [ ] Code follows the guidelines for [Logging](./docs/backend.md#logging)
* [ ] The requirements listed in [Querying the Database Safely](./docs/backend.md#querying-the-database-safely) have been satisfied.
* [ ] Request review from a member of a different team.
* [ ] Have the Pivotal acceptance criteria been met for this change?

## References

* [Pivotal story](https://www.pivotaltracker.com/story/show/163604885) for this change
* [Related pivotal story](https://www.pivotaltracker.com/story/show/160187587)

### CODE REVIEW

File: pkg/rateengine/nonlinehaul.go [hunk 1]
```diff
@@ -75,37 +75,92 @@ func (re *RateEngine) fullUnpackCents(cwt unit.CWT, zip3 string, date time.Time)
...
+
+	zapFields := []zap.Field{
 		zap.Int("cwt", cwt.Int()),
+		zap.Int("days", daysInSIT),
+		zap.String("zip3", zip3),
+		zap.Time("date", date),
+		zap.Bool("isPPM", isPPM),
+		zap.Int("effectiveCWT", effectiveCWT.Int()),
+		zap.Int("servicesSchedule", sa.ServicesSchedule),
+		zap.Int("sitPDSchedule", sa.SITPDSchedule),
```
Comment (kimallen) [inline pkg/rateengine/nonlinehaul.go]: Yay!  Thanks for adding these logs

File: pkg/rateengine/nonlinehaul.go [hunk 2]
```diff
@@ -75,37 +75,92 @@ func (re *RateEngine) fullUnpackCents(cwt unit.CWT, zip3 string, date time.Time)
...
+		if err != nil {
+			return 0, errors.Wrapf(err, "No 210A rate found for schedule %v, %v pounds, date %v", sa.SITPDSchedule, effectiveCWT.ToPounds(), date)
+		}
+		sitTotal = sitTotal.AddCents(rate210A.RateCents)
+
+		rate225A, err := models.FetchTariff400ngItemRate(re.db, "225A", sa.ServicesSchedule, effectiveCWT.ToPounds(), date)
+		if err != nil {
+			return 0, errors.Wrapf(err, "No 225A rate found for schedule %v, %v pounds, date %v", sa.ServicesSchedule, effectiveCWT.ToPounds(), date)
+		}
+		sitTotal = sitTotal.AddCents(rate225A.RateCents)
```
Comment (kimallen) [inline pkg/rateengine/nonlinehaul.go]: Since 225A gets the LH discount applied instead of SIT discount, where does it get applied separately?  From what I'm understanding here is that the SIT discount is getting applied to 225A with the rest of the values

**Comment by kimallen** [comment]
Looks pretty good to me, but I'll let someone with a more refined eye approve.  The one thing I want to make sure about is that the LH Discount is applied to 225A while SIT Discount is applied to the other values.  Might be that I'm not finding the code that does that.

For some reason I get different numbers for the VHEIS locator than you have in your summary.

**Comment by reggieriser** [comment]
That's because I apparently didn't push up my latest changes somehow.  They're pushed now, so please review the latest commit and see if numbers match up now.  You should also see how there's a new struct returned by `SITCharge` that is the used by the caller to apply the two discount rates appropriately (it's done in the caller to match how all the rate engine methods work).

File: pkg/rateengine/nonlinehaul_test.go [hunk 3]
```diff
@@ -374,3 +378,60 @@ func (suite *RateEngineSuite) Test_CheckNonLinehaulChargeTotal() {
...
+			unit.Cents(0),
+		},
+	}
+
+	for _, testCase := range discountTestCases {
+		suite.T().Run(testCase.description, func(t *testing.T) {
+			sitComputation := SITComputation{
+				SITPart:            testCase.sitPart,
+				LinehaulPart:       testCase.linehaulPart,
+				NonDiscountedTotal: testCase.sitPart.AddCents(testCase.linehaulPart),
```
Comment (kimallen) [inline pkg/rateengine/nonlinehaul_test.go]: Do we use the NonDiscountedTotal somewhere?

**Comment by kimallen** [comment]
Looks a lot better with the latest commit pushed up 👍.  My numbers are matching now.

File: pkg/rateengine/nonlinehaul_test.go (hunk 3 above)
Comment (reggieriser) [reply pkg/rateengine/nonlinehaul_test.go]: No, I just added it in there as a convenience if it was ever needed and to keep the caller from having to care how the parts made the whole.  We can take it out if we want.

File: pkg/rateengine/nonlinehaul.go (hunk 1 above)
Comment (Ryan-Koch) [reply pkg/rateengine/nonlinehaul.go]: Agreed, this is nice.

File: pkg/rateengine/nonlinehaul.go [hunk 4]
```diff
@@ -75,39 +97,100 @@ func (re *RateEngine) fullUnpackCents(cwt unit.CWT, zip3 string, date time.Time)
...
-// storage fees, 185A, to the total.
-func (re *RateEngine) SitCharge(cwt unit.CWT, daysInSIT int, zip3 string, date time.Time, isPPM bool) (unit.Cents, error) {
+// Note: Assumes the caller will apply any SIT discount rate as needed (no discounts applied here).
+func (re *RateEngine) SitCharge(cwt unit.CWT, daysInSIT int, zip3 string, date time.Time, isPPM bool) (SITComputation, error) {
 	if daysInSIT == 0 {
-		return 0, nil
+		return SITComputation{}, nil
 	} else if daysInSIT < 0 {
-		return 0, errors.New("requested SitCharge for negative days in SIT")
+		return SITComputation{}, errors.New("requested SitCharge for negative days in SIT")
```
Comment (Ryan-Koch) [inline pkg/rateengine/nonlinehaul.go]: Two questions on this error condition (I'm not saying we need to hold things up over them):

-What sort of error cause bubbles up to the handler if we were to hit this?

-Do we have a test triggering this?

Comment (reggieriser) [reply pkg/rateengine/nonlinehaul.go]: The caller (which may be a handler directly or indirectly) gets this general Go `error`.  I didn't follow all places where the error is handled, but it appears it usually gets back to the handler with the cause noted here.  It looks like most handlers probably translate this into a 500 error ultimately.  That may not be ideal in some scenarios, but since this story didn't change the existing behavior in that respect, I'd rather we consider that in the error handling epic that's separately being addressed.

I did go ahead and add test cases for zero days in SIT and a negative days in SIT.

**Comment by mtorres253** [comment]
LGTM @reggieriser
